Add experimental features for Fx 122: shadowrootmode, async clipboard (#31612)
mdn/content: files/en-us/mozilla/firefox/releases/122/index.md

These features are newly shipped in Firefox 122 but are disabled by default. To experiment with them, search for the appropriate preference on the `about:config` page and set it to `true`. You can find more such features on the [Experimental features](/en-US/docs/Mozilla/Firefox/Experimental_features) page.

- Declarative shadow DOM: `dom.webcomponents.shadowdom.declarative.enabled`
  - : The {{htmlelement("template")}} element now supports a `shadowrootmode` attribute which can be set to either `open` or `closed`, the same values as the `mode` option of {{domxref("Element.attachShadow()", "attachShadow()")}} method. It allows the creation of a shadow DOM subtree declaratively. ([Firefox bug 1712140](https://bugzil.la/1712140))
- Popover API: `dom.element.popover.enabled`

>   - : Displaying popovers on top of the page content is now supported via HTML attributes or JavaScript API, including styling with the new CSS [`:popover-open`](/en-US/docs/Web/CSS/:popover-open) pseudo-class and extended support for [`::backdrop`](/en-US/docs/Web/CSS/::backdrop) pseudo-element. See more in [Popover API](/en-US/docs/Web/API/Popover_API) reference. ([Firefox bug 1823757](https://bugzil.la/1823757))

Author: @dletorey please have a look at the popover description if I got it right. Thanks!

Reviewer: This looks great to me 🎉

- Clipboard read and write: `dom.events.asyncClipboard.clipboardItem`, `dom.events.asyncClipboard.readText` and `dom.events.asyncClipboard.writeText`

>   - : The remaining part of the async {{domxref('Clipboard API')}} is now shipped behind the flag: {{domxref('ClipboardItem')}} interface and [`read()`](/en-US/docs/Web/API/Clipboard/read), [`readText()`](/en-US/docs/Web/API/Clipboard/readText), and [`write()`](/en-US/docs/Web/API/Clipboard/write) methods. A paste context menu will appear for the user to confirm when attempting to read clipboard data not provided by the same-origin page. ([Firefox bug 1809106](https://bugzil.la/1809106))

Author: @hamishwillee please have a look at the clipboard description if I got it right. Thanks!

Reviewer 2: I think Hamish is on vacation until 26th Jan, FYI.

These features are newly shipped in Firefox 122 but are disabled by default. To experiment with them, search for the appropriate preference on the `about:config` page and set it to `true`. You can find more such features on the [Experimental features](/en-US/docs/Mozilla/Firefox/Experimental_features) page.

- Declarative shadow DOM: `dom.webcomponents.shadowdom.declarative.enabled`
  - : The {{htmlelement("template")}} element now supports a `shadowrootmode` attribute which can be set to either `open` or `closed`, the same values as the `mode` option of {{domxref("Element.attachShadow()", "attachShadow()")}} method. It allows the creation of a shadow DOM subtree declaratively. ([Firefox bug 1712140](https://bugzil.la/1712140))
- Popover API: `dom.element.popover.enabled`

>   - : Displaying popovers on top of the page content is now supported via HTML attributes or JavaScript API, including styling with the new CSS [`:popover-open`](/en-US/docs/Web/CSS/:popover-open) pseudo-class and extended support for [`::backdrop`](/en-US/docs/Web/CSS/::backdrop) pseudo-element. See more in [Popover API](/en-US/docs/Web/API/Popover_API) reference. ([Firefox bug 1823757](https://bugzil.la/1823757))

Reviewer 2: Minor prose suggestion ("new" doesn't age well, usually) what do you think?

```suggestion
  - : Displaying popovers on top of page content is now supported via HTML attributes or JavaScript API, including styling with the CSS [`:popover-open`](/en-US/docs/Web/CSS/:popover-open) pseudo-class and extended support for the [`::backdrop`](/en-US/docs/Web/CSS/::backdrop) pseudo-element. See the [Popover API](/en-US/docs/Web/API/Popover_API) reference for more details. ([Firefox bug 1823757](https://bugzil.la/1823757))
```

- Clipboard read and write: `dom.events.asyncClipboard.clipboardItem`, `dom.events.asyncClipboard.readText` and `dom.events.asyncClipboard.writeText`

>   - : The remaining part of the async {{domxref('Clipboard API')}} is now shipped behind the flag: {{domxref('ClipboardItem')}} interface and [`read()`](/en-US/docs/Web/API/Clipboard/read), [`readText()`](/en-US/docs/Web/API/Clipboard/readText), and [`write()`](/en-US/docs/Web/API/Clipboard/write) methods. A paste context menu will appear for the user to confirm when attempting to read clipboard data not provided by the same-origin page. ([Firefox bug 1809106](https://bugzil.la/1809106))

Reviewer 2: Is this correct?

```suggestion
  - : The async {{domxref('Clipboard API')}} is now fully supported, including [`read()`](/en-US/docs/Web/API/Clipboard/read), [`readText()`](/en-US/docs/Web/API/Clipboard/readText), and [`write()`](/en-US/docs/Web/API/Clipboard/write) methods and the {{domxref('ClipboardItem')}} interface. A paste context menu will appear for the user to confirm when reading clipboard data not provided by the same-origin page. ([Firefox bug 1809106](https://bugzil.la/1809106))
```

These features are newly shipped in Firefox 122 but are disabled by default. To experiment with them, search for the appropriate preference on the `about:config` page and set it to `true`. You can find more such features on the [Experimental features](/en-US/docs/Mozilla/Firefox/Experimental_features) page.

- Declarative shadow DOM: `dom.webcomponents.shadowdom.declarative.enabled`

>   The {{htmlelement("template")}} element now supports a `shadowrootmode` attribute which can be set to either `open` or `closed`, the same values as the `mode` option of {{domxref("Element.attachShadow()", "attachShadow()")}} method. It allows the creation of a shadow DOM subtree declaratively. ([Firefox bug 1712140](https://bugzil.la/1712140))

Reviewer 2: ```suggestion
  The {{htmlelement("template")}} element now supports a `shadowrootmode` attribute which can be set to either `open` or `closed`, the same values as the `mode` option of the {{domxref("Element.attachShadow()", "attachShadow()")}} method. It allows the creation of a shadow DOM subtree declaratively. ([Firefox bug 1712140](https://bugzil.la/1712140))
```
